cmd/tailscale/cli/up.go

```diff
@@ -1102,3 +1111,96 @@ func anyPeerAdvertisingRoutes(st *ipnstate.Status) bool {
 	}
 	return false
 }
+
+// resolveAuthKey either returns v unchanged (in the common case)
+// or, if it starts with "oauth:" parses it as one of:
+//
+//	oauth2:CLIENT_ID:CLIENT_SECRET?ephemeral=BOOL&tags=foo,bar&preauthorized=BOOL
+//	oauth2:CLIENT_ID:CLIENT_SECRET:BASE_URL?...
+//
+// and does the OAuth2 dance to get and return an authkey.
+//
+// If the BASE_URL argument is not provided, it defaults to https://api.tailscale.com.
+func resolveAuthKey(ctx context.Context, v string) (string, error) {
+	suff, ok := strings.CutPrefix(v, "oauth:")
+	if !ok {
+		return v, nil
+	}
+	if !envknob.Bool("TS_EXPERIMENT_OAUTH_AUTHKEY") {
+		return "", errors.New("oauth authkeys are in experimental status")
+	}
```

Comment:
I figured this lets us get this into 1.40.0 with minimal testing. We can use it ourselves in this release and remove this limitation later.

Reply:
What do you think about a `tailscale debug get-authkey` to get an authkey in one part of the script and pass it to `tailscale up --authkey=...` later?
 
I bring it up because one is likely to run `tailscale up` in the same Actions step where `tailscaled` is running, which means the client-id and secret are sitting in environment variables through the whole run.

With a `tailscale debug get-authkey`, the secrets would only be passed to an Action step which outputs the authkey string, they wouldn't be present in the step where tailscaled runs.

Reply:
Hm. But `tailscale debug get-authkey` isn't where we want to end up CLI-wise, though.

Reply:
Why do they need to be in the same `run:` step?

Reply:
I'm looking at https://github.com/tailscale/github-action/blob/main/action.yml#L57-L68 and can't that be two steps?

Reply:
We could definitely make our Action do so in two steps.

I'm more thinking of how other people would use it, where it is a very common pattern to:
```
tailscaled &
tailscale up --authkey=...
```

They'd end up with the client-id and secret in environment variables in tailscaled.
Which isn't immediately a problem, just less safe than them not being present.

Reply:
We could document best practice somewhere?

I'd prefer to keep it part of `up --authkey=` rather than adding a `debug` thing that isn't.